core/backend/src/CodeService.ts

```diff
@@ -0,0 +1,521 @@
+  */
+  export function getSequence(name: string): CodeSequence {
+    const seq = codeSequences.get(name);
+    if (!seq)
+      throw new Error("SequenceNotFound", -1, `code sequence ${name} not found`);
+    return seq;
+  }
+
+  /**
+   * Turn a `CodePops` for the briefcase of this CodeService into a `ScopeAndSpec` object for use with a CodeService.
+   * This is necessary because the `spec` member of `CodeProps` refers to the id of a code spec in the iModel, and
+   * the `scope` member refers to the element Id of the scope element in the iModel. This helper function
+   * converts the spec Id to the spec name and looks up the `FederationGuid` of the scope element.
+   */
+  export function makeScopeAndSpec(briefcase: BriefcaseDb, props: CodeProps): CodeService.ScopeAndSpec {
+    const scope = briefcase.elements.getElementProps(props.scope).federationGuid;
+    if (undefined === scope)
+      throw new CodeService.Error("MissingGuid", IModelStatus.InvalidCode, "code scope element has no federationGuid");
+
+    return { scope, spec: briefcase.codeSpecs.getById(props.spec).name };
+  }
+
+  /** Turn a `CodeProps` and  `ProposedCodeProps` into a `ProposedCode` for use with a CodeService.
+   * @see [[makeScopeAndSpec]] for explanation of why this is necessary.
+   */
+  export function makeProposedCode(arg: CodeService.MakeProposedCodeArgs): CodeService.ProposedCode {
+    return {
+      ...arg.props,
+      value: arg.code.value,
+      ...makeScopeAndSpec(arg.briefcase, arg.code),
+    };
+  }
+
+  /** The name of a code spec */
+  export type CodeSpecName = string;
+
+  /** The name that identifies the "originator" of a code. Usually this is the Guid of the iModel from which a code was added,
+   * but can also be used to identify a system or type from an external code service. */
+  export type CodeOriginName = string;
+
+  /** The name that identifies the "author" of a code. Generally, this is intended to be the name of a person or group that helps identify the purpose of the code. */
+  export type AuthorName = string;
+
+  /** The value for a code. */
+  export type CodeValue = string;
+
+  /** The guid for a code. This identifies the real-world entity associated with the code. */
+  export type CodeGuid = GuidString;
+
+  /** The guid of the scope for a code. This identifies the real-world entity that provides the uniqueness scope for code values. */
+  export type ScopeGuid = GuidString;
+
+  /** An optional number associated with a code that may be used for "status" information. Values must be defined by applications. */
+  export type CodeState = number;
+
+  /** The return status of an iterator function. The value "stop" causes the iteration to terminate. */
+  export type IteratorReturn = void | "stop";
+
+  /** An iterator function over codes in a code index. It is called with the Guid of a each code. */
+  export type CodeIterator = (guid: GuidString) => IteratorReturn;
```

Comment:
I'm not suggesting you do anything about this but the name "iterator" has a standardized meaning in javascript as an object that supplies values, not a function that receives values.

Reply:
changed to "Iteration"